File: src/nl/rubensten/texifyidea/structure/LatexStructureViewModel.java
package nl.rubensten.texifyidea.structure;

import com.intellij.ide.structureView.StructureViewModel.ElementInfoProvider;
import com.intellij.ide.structureView.StructureViewModelBase;
import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.util.treeView.smartTree.Filter;
import com.intellij.ide.util.treeView.smartTree.Sorter;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiFile;
import nl.rubensten.texifyidea.structure.filter.LatexLabelFilter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author Ruben Schellekens
 */
public class LatexStructureViewModel extends StructureViewModelBase implements ElementInfoProvider {

    public static final Sorter[] SORTERS = new Sorter[] { Sorter.ALPHA_SORTER };
    public static final Filter[] FILTERS = new Filter[] {
            new LatexLabelFilter()
    };

    public LatexStructureViewModel(@NotNull PsiFile psiFile, @Nullable Editor editor) {
        super(psiFile, editor, new LatexStructureViewElement(psiFile));
        getFilters();
    }

    @NotNull
    @Override
    public Sorter[] getSorters() {
        return SORTERS;
    }

    @NotNull
    @Override
    public Filter[] getFilters() {
        return FILTERS;
    }

    @Override
    public boolean isAlwaysShowsPlus(StructureViewTreeElement structureViewTreeElement) {
        return false;
    }

    @Override
    public boolean isAlwaysLeaf(StructureViewTreeElement structureViewTreeElement) {
        return false;
    }
}

File: src/nl/hannahsten/texifyidea/structure/latex/LatexStructureViewModel.kt
package nl.hannahsten.texifyidea.structure.latex

import com.intellij.ide.structureView.StructureViewModel.ElementInfoProvider
import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.structureView.TextEditorBasedStructureViewModel
import com.intellij.ide.util.treeView.smartTree.Sorter
import com.intellij.openapi.editor.Editor
import com.intellij.psi.PsiFile
import nl.hannahsten.texifyidea.psi.LatexCommands
import nl.hannahsten.texifyidea.structure.filter.*

/**
 * @author Hannah Schellekens
 */
class LatexStructureViewModel(
    psiFile: PsiFile,
    editor: Editor?
) : TextEditorBasedStructureViewModel(editor, psiFile), ElementInfoProvider {

    companion object {

        private val sorterArray = arrayOf(Sorter.ALPHA_SORTER)
        private val filterArray = arrayOf(
            IncludesFilter(),
            SectionFilter(),
            CommandDefinitionFilter(),
            LabelFilter(),
            BibitemFilter()
        )
    }

    override fun getRoot() = LatexStructureViewElement(psiFile)

    override fun getSorters() = sorterArray

    override fun getFilters() = filterArray

    override fun isAlwaysShowsPlus(structureViewTreeElement: StructureViewTreeElement) = false

    override fun isAlwaysLeaf(structureViewTreeElement: StructureViewTreeElement) = false

    override fun getSuitableClasses(): Array<Class<LatexCommands>> = arrayOf(
        LatexCommands::class.java
    )
}